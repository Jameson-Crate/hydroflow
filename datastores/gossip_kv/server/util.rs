use gossip_protocol::{ClientRequest, Key};
use hydroflow::DemuxEnum;

/// Convenience enum to represent a client request with the address of the client. Makes it
/// possible to use `demux_enum` in the surface syntax.
#[derive(Debug, DemuxEnum)]
pub enum ClientRequestWithAddress<A> {
    /// A get request with the key and the address of the client.
    Get { key: Key, addr: A },
    /// A set request with the key, value and the address of the client.
    Set { key: Key, value: String, addr: A },
    /// A delete request with the key and the address of the client.
    Delete { key: Key, addr: A },
}

impl<A> ClientRequestWithAddress<A> {
    /// Create a `ClientRequestWithAddress` from a `ClientRequest` and an address.
    pub fn from_request_and_address(request: ClientRequest, addr: A) -> Self {
        match request {
            ClientRequest::Get { key } => Self::Get { key, addr },
            ClientRequest::Set { key, value } => Self::Set { key, value, addr },
            ClientRequest::Delete { key } => Self::Delete { key, addr },
        }
    }
}
