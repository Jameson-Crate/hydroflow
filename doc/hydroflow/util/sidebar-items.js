window.SIDEBAR_ITEMS = {"fn":[["bind_local_udp_socket",""],["bind_udp_socket",""],["bind_udp_socket_addr",""],["collect_ready","Collects the immediately available items into the stream."],["deserialize_msg",""],["deserialize_simple",""],["ipv4_resolve",""],["recv_into","Receives available items in an `UnboundedReceiverStream` into a `FromIterator` collection."],["serialize_msg",""],["udp_bytes",""],["udp_framed",""],["udp_lines",""],["unbounded_channel",""]],"struct":[["CollectReady",""]],"type":[["UdpFramedSink",""],["UdpFramedStream",""],["UdpSink",""],["UdpStream",""]]};