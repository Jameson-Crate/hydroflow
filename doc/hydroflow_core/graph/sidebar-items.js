window.SIDEBAR_ITEMS = {"enum":[["Color",""],["Node",""]],"fn":[["node_color",""]],"mod":[["flat_graph",""],["ops",""],["partitioned_graph",""]],"struct":[["NodeId","ID to identify a node (operator or handoff) in both [`flat_graph::FlatGraph`] and [`partitioned_graph::PartitionedGraph`]."],["SubgraphId","ID to identify a subgraph in [`partitioned_graph::PartitionedGraph`]."]],"type":[["EdgePort",""],["EdgePortRef",""]]};