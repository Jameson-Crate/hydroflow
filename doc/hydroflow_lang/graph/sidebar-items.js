window.SIDEBAR_ITEMS = {"enum":["Color","Node","PortIndexValue"],"fn":["get_operator_generics","node_color"],"mod":["flat_to_partitioned","graph_algorithms","ops","serde_graph"],"struct":["DiMulGraph","FlatGraph","FlatGraphBuilder","FlatGraphExploded","GraphEdgeId","GraphNodeId","GraphSubgraphId","OpInstGenerics","OperatorInstance","PartitionedGraph"]};