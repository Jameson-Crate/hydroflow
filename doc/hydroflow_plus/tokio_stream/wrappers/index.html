<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Wrappers for Tokio types that implement `Stream`."><title>hydroflow_plus::tokio_stream::wrappers - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-9ee3a5e31a2afa3e.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydroflow_plus" data-themes="" data-resource-suffix="" data-rustdoc-version="1.75.0-nightly (cd674d617 2023-10-24)" data-channel="nightly" data-search-js="search-8fbf244ebcf71464.js" data-settings-js="settings-74424d7eec62a23e.js" ><script src="../../../static.files/storage-fec3eaa3851e447d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-5f34af1a0ee6bacd.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-5d8b3c7633ad77ba.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydroflow_plus/index.html">hydroflow_plus</a><span class="version">0.5.0</span></h2></div><h2 class="location"><a href="#">Module wrappers</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li></ul></section><h2><a href="../index.html">In hydroflow_plus::tokio_stream</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">hydroflow_plus</a>::<wbr><a href="../index.html">tokio_stream</a>::<wbr><a class="mod" href="#">wrappers</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Wrappers for Tokio types that implement <code>Stream</code>.</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="errors/index.html" title="mod hydroflow_plus::tokio_stream::wrappers::errors">errors</a></div><div class="desc docblock-short">Error types for the wrappers.</div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.BroadcastStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::BroadcastStream">BroadcastStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/sync/broadcast/struct.Receiver.html" title="struct hydroflow_plus::tokio::sync::broadcast::Receiver"><code>tokio::sync::broadcast::Receiver</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.IntervalStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::IntervalStream">IntervalStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/time/struct.Interval.html" title="struct hydroflow_plus::tokio::time::Interval"><code>Interval</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.LinesStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::LinesStream">LinesStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/io/struct.Lines.html" title="struct hydroflow_plus::tokio::io::Lines"><code>tokio::io::Lines</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.ReceiverStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::ReceiverStream">ReceiverStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/sync/mpsc/struct.Receiver.html" title="struct hydroflow_plus::tokio::sync::mpsc::Receiver"><code>tokio::sync::mpsc::Receiver</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.SplitStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::SplitStream">SplitStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/io/struct.Split.html" title="struct hydroflow_plus::tokio::io::Split"><code>tokio::io::Split</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.UnboundedReceiverStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::UnboundedReceiverStream">UnboundedReceiverStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/sync/mpsc/struct.UnboundedReceiver.html" title="struct hydroflow_plus::tokio::sync::mpsc::UnboundedReceiver"><code>tokio::sync::mpsc::UnboundedReceiver</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.WatchStream.html" title="struct hydroflow_plus::tokio_stream::wrappers::WatchStream">WatchStream</a></div><div class="desc docblock-short">A wrapper around <a href="../../tokio/sync/watch/struct.Receiver.html" title="struct hydroflow_plus::tokio::sync::watch::Receiver"><code>tokio::sync::watch::Receiver</code></a> that implements <a href="../trait.Stream.html" title="trait hydroflow_plus::tokio_stream::Stream"><code>Stream</code></a>.</div></li></ul></section></div></main></body></html>