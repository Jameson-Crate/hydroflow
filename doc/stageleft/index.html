<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `stageleft` crate."><title>stageleft - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="stageleft" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.0-nightly (bf3c6c5be 2024-02-01)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-4c98445ec4002617.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../stageleft/index.html">stageleft</a><span class="version">0.1.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../stageleft/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">stageleft</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/stageleft/lib.rs.html#1-248">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="internal/index.html" title="mod stageleft::internal">internal</a></div></li><li><div class="item-name"><a class="mod" href="runtime_support/index.html" title="mod stageleft::runtime_support">runtime_support</a></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.PATH_SEPARATOR.html" title="macro stageleft::PATH_SEPARATOR">PATH_SEPARATOR</a></div></li><li><div class="item-name"><a class="macro" href="macro.q.html" title="macro stageleft::q">q</a></div><div class="desc docblock-short">Creates a quoted expression, which can be typechecked but has its AST serialized
until it is spliced into a staged entrypoint. Returns a value which implements
<code>Quoted&lt;T&gt;</code>, where <code>T</code> is the type of the expression, and also may implement
<code>IntoQuotedMut&lt;T&gt;</code> is the values it captures would be safe to capture inside an
<code>FnMut</code> context.</div></li><li><div class="item-name"><a class="macro" href="macro.quse_fn.html" title="macro stageleft::quse_fn">quse_fn</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_crate.html" title="macro stageleft::stageleft_crate">stageleft_crate</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_macro_crate.html" title="macro stageleft::stageleft_macro_crate">stageleft_macro_crate</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_no_entry_crate.html" title="macro stageleft::stageleft_no_entry_crate">stageleft_no_entry_crate</a></div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.RuntimeData.html" title="struct stageleft::RuntimeData">RuntimeData</a></div><div class="desc docblock-short">Represents a piece of data that will be passed into the generated code</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.IntoQuotedMut.html" title="trait stageleft::IntoQuotedMut">IntoQuotedMut</a></div></li><li><div class="item-name"><a class="trait" href="trait.IntoQuotedOnce.html" title="trait stageleft::IntoQuotedOnce">IntoQuotedOnce</a></div></li><li><div class="item-name"><a class="trait" href="trait.Quoted.html" title="trait stageleft::Quoted">Quoted</a></div></li><li><div class="item-name"><a class="trait" href="trait.QuotedContext.html" title="trait stageleft::QuotedContext">QuotedContext</a></div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.entry.html" title="attr stageleft::entry">entry</a></div><div class="desc docblock-short">Defines an entrypoint for staged code, which will be available as a proc macro.
The entrypoint must be a function that returns <code>impl Quoted&lt;T&gt;</code> for some type <code>T</code>.</div></li><li><div class="item-name"><a class="attr" href="attr.runtime.html" title="attr stageleft::runtime">runtime</a></div><div class="desc docblock-short">Marks a chunk of code as being runtime-only, which means that no staged code in its crate can depend on it.
Code behind this attribute is allowed to use staged entrypoints defined in the same crate.</div></li></ul></section></div></main></body></html>