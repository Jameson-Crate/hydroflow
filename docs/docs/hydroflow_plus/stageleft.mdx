---
title: Stageleft
sidebar_position: 4
---

import StageleftDocs from '../../../stageleft/README.md'

<StageleftDocs/>
