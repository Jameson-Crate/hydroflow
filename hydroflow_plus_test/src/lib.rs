stageleft::stageleft_crate!(hydroflow_plus_test_macro);

use hydroflow_plus::futures::stream::Stream;
use hydroflow_plus::scheduled::graph::Hydroflow;
use hydroflow_plus::tokio::sync::mpsc::UnboundedSender;
use hydroflow_plus::tokio_stream::wrappers::UnboundedReceiverStream;
use hydroflow_plus::HfBuilder;
use stageleft::{q, Quoted, RuntimeData};

#[stageleft::entry(UnboundedReceiverStream<u32>)]
pub fn teed_join<'a, S: Stream<Item = u32> + Unpin + 'a>(
    graph: &'a HfBuilder<'a>,
    input_stream: RuntimeData<S>,
    output: RuntimeData<&'a UnboundedSender<u32>>,
    send_twice: bool,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let source = graph.source_stream(q!(input_stream));
    let map1 = source.map(q!(|v| (v + 1, ())));
    let map2 = source.map(q!(|v| (v - 1, ())));

    let joined = map1.join(&map2).map(q!(|t| t.0));

    joined.for_each(q!(|v| {
        output.send(v).unwrap();
    }));

    if send_twice {
        joined.for_each(q!(|v| {
            output.send(v).unwrap();
        }));
    }

    graph.build()
}

#[stageleft::entry]
pub fn chat_app<'a>(
    graph: &'a HfBuilder<'a>,
    users_stream: RuntimeData<UnboundedReceiverStream<u32>>,
    messages: RuntimeData<UnboundedReceiverStream<String>>,
    output: RuntimeData<&'a UnboundedSender<(u32, String)>>,
    replay_messages: bool,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let users = graph.source_stream(users_stream).persist();
    let mut messages = graph.source_stream(messages);
    if replay_messages {
        messages = messages.persist();
    }

    let mut joined = users.cross_product(&messages);
    if replay_messages {
        joined = joined.delta();
    }

    joined.for_each(q!(|t| {
        output.send(t).unwrap();
    }));

    graph.build()
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use hydroflow_plus::assert_graphvis_snapshots;
    use hydroflow_plus::util::collect_ready;

    use super::*;

    #[test]
    fn test_teed_join() {
        let (in_send, input) = hydroflow_plus::util::unbounded_channel::<u32>();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel::<u32>();

        let mut joined = teed_join!(input, &out, false);
        assert_graphvis_snapshots!(joined);

        in_send.send(1).unwrap();
        in_send.send(2).unwrap();
        in_send.send(3).unwrap();
        in_send.send(4).unwrap();

        joined.run_tick();

        assert_eq!(&*collect_ready::<Vec<_>, _>(&mut out_recv), &[2, 3]);
    }

    #[test]
    fn test_teed_join_twice() {
        let (in_send, input) = hydroflow_plus::util::unbounded_channel::<u32>();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel::<u32>();

        let mut joined = teed_join!(input, &out, true);
        assert_graphvis_snapshots!(joined);

        in_send.send(1).unwrap();
        in_send.send(2).unwrap();
        in_send.send(3).unwrap();
        in_send.send(4).unwrap();

        joined.run_tick();

        assert_eq!(&*collect_ready::<Vec<_>, _>(&mut out_recv), &[2, 2, 3, 3]);
    }

    #[test]
    fn test_chat_app_no_replay() {
        let (users_send, users) = hydroflow_plus::util::unbounded_channel::<u32>();
        let (messages_send, messages) = hydroflow_plus::util::unbounded_channel::<String>();
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel::<(u32, String)>();

        let mut chat_server = chat_app!(users, messages, &out, false);
        assert_graphvis_snapshots!(chat_server);

        users_send.send(1).unwrap();
        users_send.send(2).unwrap();

        messages_send.send("hello".to_string()).unwrap();
        messages_send.send("world".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (1, "hello".to_string()),
                (2, "hello".to_string()),
                (1, "world".to_string()),
                (2, "world".to_string())
            ]
        );

        users_send.send(3).unwrap();

        messages_send.send("goodbye".to_string()).unwrap();

        chat_server.run_tick();

        assert_eq!(
            &*collect_ready::<Vec<_>, _>(&mut out_recv),
            &[
                (1, "goodbye".to_string()),
                (2, "goodbye".to_string()),
                (3, "goodbye".to_string())
            ]
        );
    }

    // Fails due to Hydroflow bug
    // #[test]
    // fn test_chat_app_replay() {
    //     let (users_send, users) = hydroflow_plus::util::unbounded_channel::<u32>();
    //     let (messages_send, messages) = hydroflow_plus::util::unbounded_channel::<String>();
    //     let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel::<(u32, String)>();

    //     let mut chat_server = chat_app!(users, messages, &out, true);
    //     assert_graphvis_snapshots!(chat_server);

    //     users_send.send(1).unwrap();
    //     users_send.send(2).unwrap();

    //     messages_send.send("hello".to_string()).unwrap();
    //     messages_send.send("world".to_string()).unwrap();

    //     chat_server.run_tick();

    //     assert_eq!(
    //         &*collect_ready::<Vec<_>, _>(&mut out_recv),
    //         &[
    //             (1, "hello".to_string()),
    //             (2, "hello".to_string()),
    //             (1, "world".to_string()),
    //             (2, "world".to_string())
    //         ]
    //     );

    //     users_send.send(3).unwrap();

    //     messages_send.send("goodbye".to_string()).unwrap();

    //     chat_server.run_tick();

    //     assert_eq!(
    //         &*collect_ready::<Vec<_>, _>(&mut out_recv),
    //         &[
    //             (3, "hello".to_string()),
    //             (3, "world".to_string()),
    //             (1, "goodbye".to_string()),
    //             (2, "goodbye".to_string()),
    //             (3, "goodbye".to_string())
    //         ]
    //     );
    // }
}
