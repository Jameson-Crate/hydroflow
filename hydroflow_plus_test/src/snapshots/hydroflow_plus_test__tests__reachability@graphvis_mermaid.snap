---
source: hydroflow_plus_test/src/lib.rs
expression: "reachability.meta_graph().unwrap().to_mermaid(&Default::default())"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1[\"(1v1) <code>source_stream(roots)</code>"/]:::pullClass
2v1[\"(2v1) <code>tee()</code>"/]:::pullClass
3v1[\"(3v1) <code>source_stream(edges)</code>"/]:::pullClass
4v1[\"(4v1) <code>tee()</code>"/]:::pullClass
5v1[\"(5v1) <code>tee()</code>"/]:::pullClass
6v1[\"(6v1) <code>union()</code>"/]:::pullClass
7v1[/"(7v1) <code>tee()</code>"\]:::pushClass
8v1[/"<div style=text-align:center>(8v1)</div> <code>map({<br>    use crate::__staged::*;<br>    |r| (r, ())<br>})</code>"\]:::pushClass
9v1[\"(9v1) <code>tee()</code>"/]:::pullClass
10v1[\"(10v1) <code>join::&lt;'tick, 'tick&gt;()</code>"/]:::pullClass
11v1[\"(11v1) <code>tee()</code>"/]:::pullClass
12v1[\"<div style=text-align:center>(12v1)</div> <code>map({<br>    use crate::__staged::*;<br>    |(_from, (_, to))| to<br>})</code>"/]:::pullClass
13v1[\"(13v1) <code>tee()</code>"/]:::pullClass
14v1[/"(14v1) <code>unique::&lt;'tick&gt;()</code>"\]:::pushClass
15v1[/"(15v1) <code>tee()</code>"\]:::pushClass
16v1[/"<div style=text-align:center>(16v1)</div> <code>for_each({<br>    use crate::__staged::*;<br>    let reached_out = reached_out;<br>    |v| {<br>        reached_out.send(v).unwrap();<br>    }<br>})</code>"\]:::pushClass
17v1["(17v1) <code>handoff</code>"]:::otherClass
1v1-->2v1
3v1-->4v1
6v1-->7v1
2v1-->|0|6v1
5v1-->|1|6v1
8v1-->17v1
7v1-->8v1
10v1-->11v1
9v1-->|0|10v1
4v1-->|1|10v1
12v1-->13v1
11v1-->12v1
13v1-->5v1
14v1-->15v1
7v1-->14v1
15v1-->16v1
17v1-->9v1
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    17v1
    1v1
    2v1
    3v1
    4v1
    9v1
    10v1
    11v1
    12v1
    13v1
    5v1
    6v1
    7v1
    8v1
    14v1
    15v1
    16v1
    subgraph sg_1v1_var_stream_0 ["var <tt>stream_0</tt>"]
        1v1
        2v1
    end
    subgraph sg_1v1_var_stream_1 ["var <tt>stream_1</tt>"]
        3v1
        4v1
    end
    subgraph sg_1v1_var_stream_2 ["var <tt>stream_2</tt>"]
        5v1
    end
    subgraph sg_1v1_var_stream_3 ["var <tt>stream_3</tt>"]
        6v1
        7v1
    end
    subgraph sg_1v1_var_stream_4 ["var <tt>stream_4</tt>"]
        8v1
        9v1
    end
    subgraph sg_1v1_var_stream_5 ["var <tt>stream_5</tt>"]
        10v1
        11v1
    end
    subgraph sg_1v1_var_stream_6 ["var <tt>stream_6</tt>"]
        12v1
        13v1
    end
    subgraph sg_1v1_var_stream_7 ["var <tt>stream_7</tt>"]
        14v1
        15v1
    end
    subgraph sg_1v1_var_stream_8 ["var <tt>stream_8</tt>"]
        16v1
    end
end

